File: src/main/java/ticklock/entity/TicketTypeEntity.java
package ticklock.entity;

import jakarta.persistence.*;

@Entity
@Table(name = "ticket_types")
public class TicketTypeEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;

    private int price;

    private int totalSeats;

    private int remainingSeats;

    @Version
    private Long version;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "event_id")
    private EventEntity event;

    protected TicketTypeEntity() {
    }

    public TicketTypeEntity(String name, int price, int totalSeats, EventEntity event) {
        this.name = name;
        this.price = price;
        this.totalSeats = totalSeats;
        this.remainingSeats = totalSeats;
        this.event = event;
    }

    public Long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public int getPrice() {
        return price;
    }

    public int getTotalSeats() {
        return totalSeats;
    }

    public int getRemainingSeats() {
        return remainingSeats;
    }

    public Long getVersion() {
        return version;
    }

    public EventEntity getEvent() {
        return event;
    }

    public boolean hasRemainingSeats() {
        return remainingSeats > 0;
    }

    public void decreaseSeat() {
        if (remainingSeats <= 0) {
            throw new IllegalStateException("남은 좌석이 없습니다: " + name);
        }
        this.remainingSeats--;
    }
}

File: src/main/kotlin/ticklock/entity/TicketTypeEntity.kt
package ticklock.entity

import jakarta.persistence.*

@Entity
@Table(name = "ticket_types")
class TicketTypeEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    val name: String,

    val price: Int,

    val totalSeats: Int,

    var remainingSeats: Int = totalSeats,

    @Version
    val version: Long? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "event_id")
    val event: EventEntity? = null
) {
    constructor(name: String, price: Int, totalSeats: Int, event: EventEntity) : this(
        id = null,
        name = name,
        price = price,
        totalSeats = totalSeats,
        remainingSeats = totalSeats,
        version = null,
        event = event
    )
    fun hasRemainingSeats(): Boolean = remainingSeats > 0

    fun decreaseSeat() {
        if (remainingSeats <= 0) {
            throw IllegalStateException("남은 좌석이 없습니다: $name")
        }
        remainingSeats--
    }
}